Make dynamo-FSDP skip guards

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97463

Create a new GuardSource for FSDP modules, and use it
to opt out of guard installation.

Based on @awgu's work in https://github.com/pytorch/pytorch/pull/97091

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

wconstab committed:
Make dynamo-FSDP skip guards

Create a new GuardSource for FSDP modules, and use it
to opt out of guard installation.

[ghstack-poisoned]

wconstab committed:
Update on "Make dynamo-FSDP skip guards"


Create a new GuardSource for FSDP modules, and use it
to opt out of guard installation.

Based on awgu's work in https://github.com/pytorch/pytorch/pull/97091

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Make dynamo-FSDP skip guards"


Create a new GuardSource for FSDP modules, and use it
to opt out of guard installation.

Based on awgu's work in https://github.com/pytorch/pytorch/pull/97091

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/variables/builder.py:
```diff
@@ -645,17 +645,28 @@ def wrap_module(self, value: torch.nn.Module):
-            if getattr(value, "_is_fsdp_managed_module", False):
-                # Note: we can't do this assert inside FSDP constructor,
-                # since we don't know yet whether dynamo will be used
-                assert getattr(
-                    value, "_fsdp_use_orig_params", False
-                ), "Dynamo only supports FSDP with use_orig_params=True"
+            return UnspecializedNNModuleVariable(
+                value, guards=self.make_guards(GuardBuilder.TYPE_MATCH)
+            )
+        elif getattr(value, "_is_fsdp_managed_module", False):
```
why is `_is_fsdp_managed_module` False here? shouldn't it be true?

voznesenskym replied:
disccused offline, tab back in

voznesenskym commented on torch/_dynamo/variables/builder.py:
```diff
@@ -645,17 +645,28 @@ def wrap_module(self, value: torch.nn.Module):
+        elif getattr(value, "_is_fsdp_managed_module", False):
+            # Note: we can't do this assert inside FSDP constructor,
+            # since we don't know yet whether dynamo will be used
+            assert getattr(
+                value, "_fsdp_use_orig_params", False
+            ), "Dynamo only supports FSDP with use_orig_params=True"
```
If it only works with True, why are we asserting that it is False? Can you add more info here?

voznesenskym commented on torch/_dynamo/variables/builder.py:
```diff
@@ -645,17 +645,28 @@ def wrap_module(self, value: torch.nn.Module):
+                value, "_fsdp_use_orig_params", False
+            ), "Dynamo only supports FSDP with use_orig_params=True"
+
+            guard_create_fns = [GuardBuilder.TYPE_MATCH]
+            if torch._dynamo.config.skip_fsdp_guards:
+                guard_create_fns = [GuardBuilder.ID_MATCH]
```
Make a top level GuardBuilder FN, always use that, put the config check inside there.

voznesenskym replied:
Discussed offline - let's add a comment about the uni-directionality of modules going from no fsdp to fsdp, that's enough for now.

voznesenskym commented on torch/_dynamo/variables/builder.py:
```diff
@@ -700,7 +711,12 @@ def wrap_literal(self, value):
+        source = self.get_source()
+
+        if (
+            source.guard_source().is_nn_module()
+            and not source.guard_source().is_fsdp_module()
```
Why are you skipping registration? This is not correct and will result in missing FQNs.

voznesenskym replied:
Also, if you get rid of this, you can get rid of alllll the `is_fsdp_module` logic and just do what https://github.com/pytorch/pytorch/pull/96905 does. Or did you find that PR missing some cases?

voznesenskym commented on torch/_dynamo/variables/nn_module.py:
```diff
@@ -664,7 +670,28 @@ def call_method(
+    """
+
+    def __init__(self, value, **kwargs):
+        super().__init__(value=value, **kwargs)
+        if torch._dynamo.config.skip_fsdp_guards:
+            self.source = FSDPNNModuleSource(self.source)
+        else:
+            self.source = NotNNModuleSource(self.source)
```
Super for this class does:

`self.source = NotNNModuleSource(self.source)`

So I think you might end up with a `NotNNModuleSource` in the chain?

I think it would be clearer to move source writing into a method so that super always calls the lowest level of the hierarchy that overrides it, like https://github.com/pytorch/pytorch/pull/96905/files#diff-890f369c8fb7fd76d79bac7fe4f8dddf126916580d0a9ee6b64dad221ba3d46eR671

wconstab commented on torch/_dynamo/variables/builder.py:
```diff
@@ -645,17 +645,28 @@ def wrap_module(self, value: torch.nn.Module):
         elif getattr(value, "_is_fsdp_managed_module", False) or issubclass(
```
fix this

wconstab committed:
Update on "Make dynamo-FSDP skip guards"


Create a new GuardSource for FSDP modules, and use it
to opt out of guard installation.

Based on awgu's work in https://github.com/pytorch/pytorch/pull/97091

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Make dynamo-FSDP skip guards"


Create a new GuardSource for FSDP modules, and use it
to opt out of guard installation.

Based on awgu's work in https://github.com/pytorch/pytorch/pull/97091

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Make dynamo-FSDP skip guards"


Create a new GuardSource for FSDP modules, and use it
to opt out of guard installation.

Based on awgu's work in https://github.com/pytorch/pytorch/pull/97091

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Make dynamo-FSDP skip guards"


Create a new GuardSource for FSDP modules, and use it
to opt out of guard installation.

Based on awgu's work in https://github.com/pytorch/pytorch/pull/97091

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Make dynamo-FSDP skip guards"


Create a new GuardSource for FSDP modules, and use it
to opt out of guard installation.

Based on awgu's work in https://github.com/pytorch/pytorch/pull/97091

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

awgu commented on torch/_dynamo/variables/builder.py:
```diff
@@ -642,21 +642,43 @@ def wrap_module(self, value: torch.nn.Module):
                 self.source, value, result
             )
-        elif getattr(value, "_is_fsdp_managed_module", False) or issubclass(
+        elif issubclass(
```
Something I wondered but did not check: Does this mean that DDP uses `UnspecializedNNModuleVariable`? Or when is this code path used?

wconstab replied:
I haven't stepped through DDP in a while but iirc this only makes the DDP wrapper module itself an UnspecializedNNModule.  I can't remember why I did this.

But DDP doesn't have the 'poison' attribute like `_is_fsdp_managed_module`, so any modules inside the DDP wrapper should fall back to being regular NNModuleVariable.

wconstab committed:
Update on "Make dynamo-FSDP skip guards"


Create a new GuardSource for FSDP modules, and use it
to opt out of guard installation.

Based on awgu's work in https://github.com/pytorch/pytorch/pull/97091

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Make dynamo-FSDP skip guards"


Create a new GuardSource for FSDP modules, and use it
to opt out of guard installation.

Based on awgu's work in https://github.com/pytorch/pytorch/pull/97091

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

awgu commented on torch/_dynamo/variables/nn_module.py:
```diff
@@ -664,7 +670,32 @@ def call_method(
+    """
+
+    def __init__(self, value, **kwargs):
+        source = kwargs.get("source", None)
+        assert source is not None
```
nit: Explain what it means if `source is None`? E.g. when can `VariableBuilder.source` be `None`?

awgu commented on torch/_guards.py:
```diff
@@ -40,19 +42,34 @@ def select(self, locals_, globals_):
+
     def is_nn_module(self) -> bool:
-        return self in (GuardSource.GLOBAL_NN_MODULE, GuardSource.LOCAL_NN_MODULE)
+        return self.is_fsdp_module() or self in (
```
nit: This is negligible overhead-wise, but you could consider swapping the order of the clauses in this `or` since `self.is_fsdp_module()` is much less likely to return `True` than the other clause.

wconstab replied:
oh.. yea, i could swap it.  I actually did that first and i put it this way bc its easier to 'see' the fsdp part than if it's at the end.

awgu commented on torch/_dynamo/variables/builder.py:
```diff
@@ -642,21 +642,43 @@ def wrap_module(self, value: torch.nn.Module):
+            # Note on FSDP guarding
+            # 1. We expect FSDP wrapping mutates an nn module irreversably (no way to de-wrap).
+            # 2. Eager FSDP already assumes (requires, but without enforcement) that users don't mutate their
+            #    model parameters/structure after FSDP wrapping, becuase FSDP wouldn't notice or update its FlatParams.
```
nit: typo
```suggestion
            #    model parameters/structure after FSDP wrapping, because FSDP wouldn't notice or update its FlatParams.
```

awgu commented on torch/_dynamo/variables/nn_module.py:
```diff
@@ -664,7 +670,32 @@ def call_method(
+    register parameters to the top-level, treat them as function inputs.
+
+    Guards behavior: if 'skip_fsdp_guards', many guards that would be installed
+    by a vanilla UnspecializedNNMouduleVariable are simply dropped, on the basis
```
nit: typo
```suggestion
    by a vanilla UnspecializedNNModuleVariable are simply dropped, on the basis
```

awgu reviewed (approved):
This looks good to me, and thanks for driving this through so quickly! I left some nits.

wconstab committed:
Update on "Make dynamo-FSDP skip guards"


Create a new GuardSource for FSDP modules, and use it
to opt out of guard installation.

Based on awgu's work in https://github.com/pytorch/pytorch/pull/97091

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge -f "flaky failure now, and got a clean CI on previous commit, this one is just cosmetic changes"
